File: Source/SlopArena/Shared/CombatMath.cpp
// Copyright SlopArena Contributors. MIT License.
// NOTE: Unreal convention — X/Y = horizontal plane, Z = up axis

#include "CombatMath.h"

bool FCombatMath::IsInCircle(const FVector& Target, const FVector& Center, float Radius)
{
	float Dx = Target.X - Center.X;
	float Dy = Target.Y - Center.Y;
	return (Dx * Dx + Dy * Dy) <= Radius * Radius;
}

bool FCombatMath::IsInCone(const FVector& Target, const FVector& Origin, const FVector& Direction, float HalfAngleRad, float Range)
{
	float Dx = Target.X - Origin.X;
	float Dy = Target.Y - Origin.Y;
	float DistSq = Dx * Dx + Dy * Dy;

	if (DistSq > Range * Range)
		return false;

	if (DistSq < 0.001f)
		return true;

	float Dist = FMath::Sqrt(DistSq);
	float Dot = (Dx * Direction.X + Dy * Direction.Y) / Dist;
	Dot = FMath::Clamp(Dot, -1.0f, 1.0f);
	float Angle = FMath::Acos(Dot);
	return Angle <= HalfAngleRad;
}

bool FCombatMath::LineIntersectsCircle(
	const FVector2D& LineStart, const FVector2D& LineEnd,
	const FVector2D& CircleCenter, float Radius)
{
	FVector2D D = LineEnd - LineStart;
	FVector2D F = LineStart - CircleCenter;

	float A = D.X * D.X + D.Y * D.Y;
	float B = 2.0f * (F.X * D.X + F.Y * D.Y);
	float C = F.X * F.X + F.Y * F.Y - Radius * Radius;

	float Discriminant = B * B - 4.0f * A * C;
	if (Discriminant < 0.0f)
		return false;

	Discriminant = FMath::Sqrt(Discriminant);
	float T1 = (-B - Discriminant) / (2.0f * A);
	float T2 = (-B + Discriminant) / (2.0f * A);

	return (T1 >= 0.0f && T1 <= 1.0f) || (T2 >= 0.0f && T2 <= 1.0f) || (T1 < 0.0f && T2 > 1.0f);
}

FVector FCombatMath::CalculateKnockback(const FVector& TargetPos, const FVector& AttackerPos, float Force, float Upward)
{
	float Dx = TargetPos.X - AttackerPos.X;
	float Dy = TargetPos.Y - AttackerPos.Y;
	float DistSq = Dx * Dx + Dy * Dy;

	if (DistSq > 0.001f)
	{
		float Dist = FMath::Sqrt(DistSq);
		return FVector((Dx / Dist) * Force, (Dy / Dist) * Force, Upward);
	}

	return FVector(0.0f, Force, Upward);
}

float FCombatMath::HorizontalDistance(const FVector& A, const FVector& B)
{
	float Dx = B.X - A.X;
	float Dy = B.Y - A.Y;
	return FMath::Sqrt(Dx * Dx + Dy * Dy);
}

FVector FCombatMath::ClampHorizontalSpeed(const FVector& Velocity, float MaxSpeed)
{
	FVector Horizontal(Velocity.X, Velocity.Y, 0.0f);
	float Speed = Horizontal.Length();
	if (Speed > MaxSpeed)
	{
		float Scale = MaxSpeed / Speed;
		return FVector(Velocity.X * Scale, Velocity.Y * Scale, Velocity.Z);
	}
	return Velocity;
}

File: src/Shared/CombatMath.cs
using System;

namespace SlopArena.Shared
{
    /// <summary>
    /// Pure math functions for combat calculations.
    /// No Godot dependencies - usable by Server and Client.
    ///
    /// All positions are in 3D space (X, Y, Z).
    /// Y is the up axis.
    /// </summary>
    public static class CombatMath
    {
        /// <summary>
        /// Check if a point is within a circle (cylinder) centered at another point.
        /// Used for AoE circle hits and projectile proximity checks.
        /// </summary>
        public static bool IsInCircle(
            float targetX, float targetZ,
            float centerX, float centerZ,
            float radius)
        {
            float dx = targetX - centerX;
            float dz = targetZ - centerZ;
            float distSq = (dx * dx) + (dz * dz);
            return distSq <= radius * radius;
        }

        /// <summary>
        /// Check if a point is within a cone originating from a position.
        /// The cone is defined by its origin, direction, angle (half-angle in radians), and range.
        /// </summary>
        public static bool IsInCone(
            float targetX, float targetZ,
            float originX, float originZ,
            float dirX, float dirZ,
            float halfAngleRad, float range)
        {
            float dx = targetX - originX;
            float dz = targetZ - originZ;
            float distSq = (dx * dx) + (dz * dz);

            if (distSq > range * range)
                return false;

            if (distSq < 0.001f)
                return true; // At origin

            float dist = MathF.Sqrt(distSq);
            float dot = ((dx * dirX) + (dz * dirZ)) / dist; // Normalized dot product

            // Clamp to avoid NaN from floating point
            dot = Math.Clamp(dot, -1f, 1f);
            float angle = MathF.Acos(dot);

            return angle <= halfAngleRad;
        }

        /// <summary>
        /// Check if a line segment (projectile path) intersects a circle (entity hitbox).
        /// Uses segment-to-circle distance check.
        /// </summary>
        public static bool LineIntersectsCircle(
            float lineStartX, float lineStartZ,
            float lineEndX, float lineEndZ,
            float circleX, float circleZ,
            float radius)
        {
            float dx = lineEndX - lineStartX;
            float dz = lineEndZ - lineStartZ;
            float fx = lineStartX - circleX;
            float fz = lineStartZ - circleZ;

            float a = (dx * dx) + (dz * dz);
            float b = 2f * ((fx * dx) + (fz * dz));
            float c = (fx * fx) + (fz * fz) - (radius * radius);

            float discriminant = (b * b) - (4f * a * c);
            if (discriminant < 0)
                return false;

            discriminant = MathF.Sqrt(discriminant);
            float t1 = (-b - discriminant) / (2f * a);
            float t2 = (-b + discriminant) / (2f * a);

            // Check if intersection is within the segment [0, 1]
            return (t1 >= 0 && t1 <= 1) || (t2 >= 0 && t2 <= 1) || (t1 < 0 && t2 > 1);
        }

        /// <summary>
        /// Calculate knockback direction from attacker to target (horizontal only).
        /// Used exclusively by pull-type abilities (Bunny Jade Hare).
        /// NOT the main knockback path — see Simulation.ApplyKnockback for the
        /// standard Smash-style formula with BaseKnockback + KnockbackGrowth.
        /// </summary>
        public static void CalculateKnockback(
            float targetX, float targetZ,
            float attackerX, float attackerZ,
            float force, float upward,
            out float kbX, out float kbY, out float kbZ)
        {
            float dx = targetX - attackerX;
            float dz = targetZ - attackerZ;
            float distSq = (dx * dx) + (dz * dz);

            if (distSq > 0.001f)
            {
                float dist = MathF.Sqrt(distSq);
                kbX = (dx / dist) * force;
                kbZ = (dz / dist) * force;
            }
            else
            {
                kbX = 0f;
                kbZ = force; // Default forward
            }

            kbY = upward;
        }

        /// <summary>
        /// Distance between two 3D points (horizontal only, XZ plane).
        /// </summary>
        public static float HorizontalDistance(float x1, float z1, float x2, float z2)
        {
            float dx = x2 - x1;
            float dz = z2 - z1;
            return MathF.Sqrt((dx * dx) + (dz * dz));
        }

        /// <summary>
        /// Compute projectile launch speed and horizontal/vertical components
        /// for a targeted throw landing at a different height than launch.
        ///
        /// Formula: v² = g·D² / (2·cos²(θ)·(D·tan(θ) - dY))
        /// where D = horizontal distance, dY = targetY - launchY (negative = lower),
        /// θ = launch angle, g = gravity.
        ///
        /// Used by both client (arc visual) and server (hitbox velocity).
        /// </summary>
        /// <param name="targetDistance">Horizontal distance in meters (D).</param>
        /// <param name="launchAngleRad">Launch angle above horizontal in radians.</param>
        /// <param name="gravity">Gravity in m/s².</param>
        /// <param name="heightOffset">Vertical displacement: targetY - launchY (negative when throwing downward).</param>
        /// <param name="speed">Launch speed magnitude.</param>
        /// <param name="hSpeed">Horizontal speed component.</param>
        /// <param name="vSpeed">Vertical speed component (upward positive).</param>
        public static void ComputeProjectileLaunch(
            float targetDistance, float launchAngleRad, float gravity, float heightOffset,
            out float speed, out float hSpeed, out float vSpeed)
        {
            float tanθ = MathF.Tan(launchAngleRad);
            float cosθ = MathF.Cos(launchAngleRad);
            float denom = 2f * cosθ * cosθ * (targetDistance * tanθ - heightOffset);
            speed = denom > 0.001f
                ? MathF.Sqrt(gravity * targetDistance * targetDistance / denom)
                : MathF.Sqrt(targetDistance * gravity);
            hSpeed = speed * cosθ;
            vSpeed = speed * MathF.Sin(launchAngleRad);
        }
    }
}
